### PR TITLE
Format DOM Nesting Warning as Diff View + An Additional Log for Stack Trace

### PR DESCRIPTION
**Pull request by sebmarkbage**
Currently we're printing parent stacks at the end of DOM nesting even with owner stacks enabled. That's because the context of parent tree is relevant for determining why two things are nested. It might not be sufficient to see the owner stack alone.

I'm trying to get rid of parent stacks and rely on more of the plain owner stacks or ideally console.createTask. These are generally better anyway since the exact line for creating the JSX is available. It also lets you find a parent stack frame that is most relevant e.g. if it's hidden inside internals.

For DOM nesting there's really only two stacks that are relevant. The creation of the parent and the creation of the child. Sometimes they're close enough to be the same thing. Such as for parents that can't have text children or when the ancestor is the direct parent created at the same place (same owner).

Sometimes they're far apart. In this case I add a second console.error within the context of the ancestor. That way the second stack trace can be used to read the stack trace for where it was created.

To preserve some parent context I now print the parent stack in a diff view format using the logic from hydration diffs. This includes some siblings and props for context.

Screenshot 2024-07-10 at 12 21 38 AM

Text Nodes:

Screenshot 2024-07-10 at 12 37 40 AM

### CODE REVIEW

File: packages/react-dom-bindings/src/client/validateDOMNesting.js [hunk 1]
```diff
@@ -7,7 +7,54 @@
...
+      serverTail: [],
+      distanceFromLeaf: distanceFromLeaf,
+    };
+    distanceFromLeaf++;
+    fiber = fiber.return;
+  }
+  if (node !== null) {
+    // Describe the node using the hydration diff logic.
+    // Replace + with - to mark ancestor and child. It's kind of arbitrary.
+    return describeDiff(node).replace(/\n\+/g, '\n-');
```
Comment (tjallingt) [inline packages/react-dom-bindings/src/client/validateDOMNesting.js:54]:
```suggestion
    return describeDiff(node).replace(/^\+/gm, '-');
```
or even
```suggestion
    return describeDiff(node).replace(/^[+-]/gm, '>');
```

Comment (sebmarkbage) [reply packages/react-dom-bindings/src/client/validateDOMNesting.js]: There's some value in having a single diff format to make formatting easier but I like the idea of using `>`. Seems simple enough.

File: packages/react-dom-bindings/src/client/validateDOMNesting.js [hunk 2]
```diff
@@ -483,28 +553,50 @@ function validateDOMNesting(
...
-        getCurrentParentStackInDev(),
+        ancestorDescription,
       );
     } else {
       // Don't transform into consoleWithStackDev here because we add a manual stack.
       // We use the parent stack here instead of the owner stack because the parent
       // stack has more useful context for nesting.
       // TODO: Format this as a linkified "diff view" with props instead of
       // a stack trace since the stack trace format is now for owner stacks.
-      console['error'](
```
Comment (gnoff) [inline packages/react-dom-bindings/src/client/validateDOMNesting.js:500]: You left the comment which says not to transform the console method but I think it’s no longer valid

File: packages/react-dom-bindings/src/client/validateDOMNesting.js [hunk 3]
```diff
@@ -483,28 +553,50 @@ function validateDOMNesting(
...
+      console.error(
```
Comment (gnoff) [inline packages/react-dom-bindings/src/client/validateDOMNesting.js:556]: Update comment

**Review by gnoff — APPROVED** [review]

**Pull request merged**